### ISSUE TITLE
MIMIC repository consolidation

### ISSUE THREAD

**Issue by briangow** [MIT-LCP/mimic-code#910]
In an effort to make the organization of the MIMIC repositories clear we want to consolidate everything under mimic-code and mimic-website.  

Goals: 
1. Transfer issues (from mimic-iv, mimic-cxr, etc) into mimic-code.
2. Update references to the transferred issues based on their new issue numbers. 
3. Merge the code (from mimic-iv, mimic-cxr, etc) into mimic-code while retaining the commit log.

This issue is specific to the consolidation of the non-website code into mimic-code.  Another issue will be created in the mimic-website repository for the website code.

**Comment by briangow** [comment]
The current plan involves merging the mimic-iv and mimic-cxr repositories into mimic-code.  Other repositories will be merged or added in the future (mimic-ed, etc).  

Below I give details around what has been tried so far and how it behaves.  **Feel free to skip the details in this comment and respond to the following comment.**  

At a high level these are the steps to be followed (note that the word "issues" below does not include pull-requests unless specifically stated):

1. Add a mimic-iii label to all open and closed issues in mimic-code
2. Using the Github GUI transfer all open and closed issues from mimic-iv to mimic-code 
3. Add a mimic-iv label to the transferred issues
4. Using the Github GUI transfer all open and closed issues from mimic-cxr to mimic-code
5. Add a mimic-cxr label to the transferred issues
6. Move all of the mimic-code files under a mimic-iii folder
7. Merge the mimic-iv files into the mimic-code directory under a mimic-iv folder by following this process:
```

git remote add -f mimic-iv https://github.com/MIT-LCP/mimic-iv.git
git fetch mimic-iv —tags
# manually resolve conflicts in .gitignore
git merge mimic-iv/master --allow-unrelated-histories

mkdir mimic-iv 
shopt -s extglob
git mv !(mimic-iii|mimic-iv) /Users/briangow/Documents/MIT/mimic/github_reorg/mimic-code/mimic-iv
git add .
git commit -m “commit msg“
git push
git remote remove mimic-iv
shopt -u extglob
```
8. Repeat step 7 for the mimic-cxr repository in place of the mimic-iv repository. 

Regarding the consolidation of the issues (steps 1-5), the following behavior is observed:
* Transferred issues are assigned a new issue number (to avoid conflicts).
* Links between the transferred issues and the pull-requests (still in their original repository) are preserved. 
* Commit messages (in the GUI and log) which reference an issue # will remain unchanged from the original statement but the link (in the GUI - the log doesn't have links) will point to the updated issue #.  For example:
  * Create a new issue in mimic-iv which gets assigned issue # 7
  * Create a commit which states "fixes # 7"
  * Create and merge a pull request (# 8) for this commit
  * Transfer issue # 7 from mimic-iv to mimic-code where it is assigned # 10 as an issue number
  * The closed pull request (still under the mimic-iv repository) will still say fixes # 7.  Clicking on its hyperlink will send you to issue # 10 on mimic-code
  * The reference to the commit message will still show in the transferred issue and will also continue to say fixes # 7 even though clicking its hyperlink will direct you to the # 10 issue which you were already on.  There will also be a line which says "closed this in mimic-iv#8" which links to the original pull request still in the mimic-iv repository. 
  
If you would like to see screenshot examples of this they can be provided.

 Instead of using the Github GUI, `hub issue transfer <issue-num> <repo-to-transfer-to>` can be used to transfer issues programmatically and appears to behave similarly (https://github.com/github/hub).

Regarding the consolidation of the code (steps 6-8).  This will produce a log with the commits from all of the repositories organized by date.  An example of this can be seen in the attached [log](https://github.com/MIT-LCP/mimic-code/files/6444026/mimic-code-consol-log.txt).  Please see https://github.com/MIT-LCP/mimic-code/tree/consol_repos for an example of the consolidated code.

**Comment by briangow** [comment]
As a summary to the details above I will address whether this achieves the goals stated at the top of this issue.

1. Transfer issues (from mimic-iv, mimic-cxr, etc) into mimic-code.
  * This works as expected, issues from mimic-iv / mimic-cxr are transferred into mimic-code and assigned new issue #'s to avoid conflicts. 
2. Update references to the transferred issues based on their new issue numbers. 
  * In the Github GUI the links are updated after transferring the issues which allows for linking between transferred issues and pull requests (which don't get transferred).  However, since there aren't links in the commit log it isn't possible to simply use the commit log to recreate history.  In the commit log, a statement such as fixes # 7 will likely be pointing to the wrong issue number since the issue numbers are changed upon transfer.
3. Merge the code (from mimic-iv, mimic-cxr, etc) into mimic-code while retaining the commit log. 
  * This works but the accuracy of the commit log isn't ideal (see above) and you can't tell from the log which repository a given commit applied to.  

@alistairewj , @tompollard , @bemoody , @SichengH , @Lucas-Mc please let me know your thoughts and any suggestions for improving this. In particular, do you feel like it is ok that we don't know which repository a given commit applies to in goal # 3 and that the referenced issue #'s in goal # 2 aren't updated in the commit log?

**Comment by briangow** [comment]
@tompollard , @alistairewj , I'd like to start transferring issues from the mimic-iv, and mimic-cxr repos to mimic-code as outlined above. With mimic-code, mimic-iv, and mimic-cxr cloned locally the following script does this by:

1. Reading all of the open and closed issues from mimic-iv / cxr
2. Transferring all of these issues to mimic-code while capturing the new number for the issue in mimic-code
3. Saving all of the new issue numbers after transfer
4. Updating all of the newly transferred issues with an appropriate label: mimic-iv / mimic-cxr

This uses the [hub](https://github.com/github/hub) package mentioned above.  

Please let me know if you have any concerns about this or if you think I should figure out how to backup the issues before starting this.

```
import os

label = 'mimic-iv'

# Get list of open issues
open_stream = os.popen('hub issue')
open_issues = open_stream.read()
closed_stream = os.popen('hub issue -s closed')
closed_issues = closed_stream.read()

# Convert stream to a list
open_list = open_issues.splitlines()
closed_list = closed_issues.splitlines()

all_list = open_list + closed_list

new_issues = []
# Transfer the open issues
for issue in all_list:
    tn = issue.split()[0].split("#")[1]
    trans_stream = os.popen('hub issue transfer ' + tn + ' mimic-code')
    # Get the new issue number from the directory you transferred to
    new_issue_num = trans_stream.read().split('/')[-1].split("\n")[0]
    # Store the new issue number so you can add a label to it later
    new_issues.append(new_issue_num)

# Save new issue numbers to a file
with open('new_issues.txt', 'w') as filehandle:
    for listitem in new_issues:
        filehandle.write('%s\n' % listitem)

# Go to mimic-code repo to add labels to transferred issues
os.chdir('../mimic-code')
for new in new_issues:
    os.system('hub issue update ' + new + ' -l ' + label)
```

**Comment by briangow** [comment]
Updated script which handles transferring old labels (including ones with spaces in their name):

```
import os
import re

label = 'mimic-iv'

# Get list of open issues
open_stream = os.popen('hub issue')
open_issues = open_stream.read()
closed_stream = os.popen('hub issue -s closed')
closed_issues = closed_stream.read()

# Convert stream to a list
open_list = open_issues.splitlines()
closed_list = closed_issues.splitlines()

all_list = open_list + closed_list

new_issues = []
# Transfer the open issues
for issue in all_list:
    tn = issue.split()[0].split("#")[1]
    lbl_stream = os.popen('hub issue show -f%l ' + tn)
    prev_lbl = lbl_stream.read()
    trans_stream = os.popen('hub issue transfer ' + tn + ' mimic-code')
    # Get the new issue number from the directory you transferred to
    new_issue_num = trans_stream.read().split('/')[-1].split("\n")[0]
    # Check to see if there was a label already
    if prev_lbl:
        old_lbl = prev_lbl
    else:
        old_lbl = 'NA'
    # Store the new issue number so you can add a label to it later
    new_issues.append(new_issue_num + "," + old_lbl)

# Save new issue numbers to a file
with open('new_issues.txt', 'w') as filehandle:
    for listitem in new_issues:
        filehandle.write('%s\n' % listitem)

# Go to mimic-code repo to add labels to transferred issues
os.chdir('../mimic-code')
for new in new_issues:
    old_present = new.split(",")[1]
    if old_present == 'NA':
        write_lbl = label
    else:
        # Parse and create a string of comma separated strings
        split_lbl = re.split('   ', old_present.strip())
        split_lbl.append(label)
        write_lbl = ','.join(['"{}"'.format(value) for value in split_lbl])
    # Add label (previous + label from above) to transferred issue
    os.system('hub issue update ' + new.split(",")[0] + ' -l ' + write_lbl)

```

**Comment by Lucas-Mc** [comment]
> ```
> all_list = open_list + closed_list
> 
> new_issues = []
> # Transfer the open issues
> for issue in all_list:
> ```

Why do you say "Transfer the open issues" but then use `all_list` instead of `open_list`?

> ```
>     tn = issue.split()[0].split("#")[1]
>     lbl_stream = os.popen('hub issue show -f%l ' + tn)
> ```

Just a side note, but you mix `"` and `'` a lot which gets confusing. I stick with `'` unless necessary but I think it's a preference thing just be consistent.

> ```
>     # Get the new issue number from the directory you transferred to
>     new_issue_num = trans_stream.read().split('/')[-1].split("\n")[0]
> ```

I know you'll probably only use this on your Mac, but for other machines if you want to use in the future, using `.split(os.sep)` instead of `.split('/')` will prevent it from breaking when the file separators switch.


> ```
>         filehandle.write('%s\n' % listitem)
> ```

I'm pretty sure `%` string formatting is deprecated now as it was a Python 1 thing, also you use `.format()` somewhere else so I would switch to that. As a heads up, f-strings are the new hipping and hopping thing in Python (introduced in 3.6) which makes this stuff a lot easier.

> ```
>         split_lbl = re.split('   ', old_present.strip())
> ```

I may be a little confused, but why can't you just do `old_present.strip().split('   ')` instead and ditch the `re` package?

**Comment by tompollard** [comment]
Thanks @briangow. The following steps sound fine to me:

> 1. Add a mimic-iii label to all open and closed issues in mimic-code
> 2. Using the Github GUI transfer all open and closed issues from mimic-iv to mimic-code
> 3. Add a mimic-iv label to the transferred issues
> 4. Using the Github GUI transfer all open and closed issues from mimic-cxr to mimic-code
> 5. Add a mimic-cxr label to the transferred issues
> 6. Move all of the mimic-code files under a mimic-iii folder
> 7. Merge the mimic-iv files into the mimic-code directory under a mimic-iv folder by following this process:
> 8. Repeat step 7 for the mimic-cxr repository in place of the mimic-iv repository.

If you've tested your script and it does what you want to do, then this is good enough for me. It sounds like Lucas has also taken a look and he is generally happy, so go ahead as far as I am concerned. The script is a one-off thing, so don't worry about minor styling things.

**Comment by briangow** [comment]
Thanks @Lucas-Mc !  I made the comment and style updates you suggested.  I tend to prefer `'` also and agree it makes sense to be consistent with that.  I thought I needed to use `re.split` when spitting on multiple characters (three spaces in this case) but that doesn't actually seem to be necessary.

**Comment by briangow** [comment]
Thanks @tompollard , sounds good, I'll start the transfer.
